core/solver/direct.cpp

```diff
@@ -0,0 +1,282 @@
+      lower_solver_{std::move(other.lower_solver_)},
+      upper_solver_{std::move(other.upper_solver_)}
+{}
+
+
+template <typename ValueType, typename IndexType>
+Direct<ValueType, IndexType>& Direct<ValueType, IndexType>::operator=(
+    const Direct& other)
+{
+    if (this != &other) {
+        EnableLinOp<Direct>::operator=(other);
+        gko::solver::EnableSolverBase<Direct, factorization_type>::operator=(
+            other);
+        const auto exec = this->get_executor();
+        lower_solver_ = other.lower_solver_->clone(exec);
+        upper_solver_ = other.upper_solver_->clone(exec);
+    }
+    return *this;
+}
+
+
+template <typename ValueType, typename IndexType>
+Direct<ValueType, IndexType>& Direct<ValueType, IndexType>::operator=(
+    Direct&& other)
+{
+    if (this != &other) {
+        EnableLinOp<Direct>::operator=(std::move(other));
+        gko::solver::EnableSolverBase<Direct, factorization_type>::operator=(
+            std::move(other));
+        const auto exec = this->get_executor();
+        lower_solver_ = std::move(other.lower_solver_);
+        upper_solver_ = std::move(other.upper_solver_);
+    }
+    return *this;
+}
+
+
+template <typename ValueType, typename IndexType>
+Direct<ValueType, IndexType>::Direct(std::shared_ptr<const Executor> exec)
+    : EnableLinOp<Direct>{exec}
+{}
+
+
+template <typename ValueType, typename IndexType>
+static std::shared_ptr<const factorization::Factorization<ValueType, IndexType>>
+generate_factorization(
+    std::shared_ptr<const LinOpFactory> factorization_factory,
+    std::shared_ptr<const LinOp> system_matrix)
+{
+    if (auto factorization = std::dynamic_pointer_cast<
+            const factorization::Factorization<ValueType, IndexType>>(
+            system_matrix)) {
+        return factorization;
+    } else {
+        return as<factorization::Factorization<ValueType, IndexType>>(
+            factorization_factory->generate(system_matrix));
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+Direct<ValueType, IndexType>::Direct(const Factory* factory,
+                                     std::shared_ptr<const LinOp> system_matrix)
+    : EnableLinOp<Direct>{factory->get_executor(), system_matrix->get_size()},
+      gko::solver::EnableSolverBase<
+          Direct, factorization::Factorization<ValueType, IndexType>>{
+          generate_factorization<ValueType, IndexType>(
+              factory->get_parameters().factorization, system_matrix)}
+{
+    using factorization::storage_type;
+    const auto lower_factory = lower_type::build().on(factory->get_executor());
+    const auto upper_factory = upper_type::build().on(factory->get_executor());
+    const auto factors = this->get_system_matrix();
+    switch (factors->get_storage_type()) {
+    case storage_type::empty:
+        // leave both solvers as nullptr
+        break;
+    case storage_type::composition:
+    case storage_type::symm_composition:
+        // TODO handle diagonal
```

Comment:
this will most likely be diagonal scaling exclusively, so I think an in-place kernel inside direct.cpp is most suitable here.

Reply:
`Factorization` already has a `get_diagonal` so at least for the composition that seems to be straight forward. Also, would there be any difference to applying a diagonal matrix (or its inverse) with an in-place kernel?

Reply:
the inverse is probably the most important thing preventing the use of the Diagonal matrix type directly. I would differentiate between compositions and combined types, because the latter need to store the diagonal index for each row.